File: RemoteSigner/Kubernetes.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Security;
using System.Security.Cryptography.X509Certificates;
using System.Threading.Tasks;
using Newtonsoft.Json;
using RemoteSigner.Log;
using RemoteSigner.Models.Kubernetes;

namespace RemoteSigner {
    public class Kubernetes {
        private const string ServiceAccountPath = "/run/secrets/kubernetes.io/serviceaccount";
        private const string KubernetesLog = "Kubernetes";

        private static JsonSerializerSettings jsonSettings;
        
        public static bool InKubernetes { get; private set; }
        public static string Namespace { get; private set; }
        public static Pod Me { get; private set; }
        public static string Hostname { get; private set; }

        private static string ServiceURL => $"https://kubernetes.default.svc/api/v1/namespaces/{Namespace}/services";
        private static string PodURL => $"https://kubernetes.default.svc/api/v1/namespaces/{Namespace}/pods";

        private static string KubeToken { get; set; }
        
        public static string KubeCA { get; private set; }
        public static X509Certificate KubeCAX509 { get; private set; }

        public static void Init() {
            Logger.Log(KubernetesLog, "Checking if running in kubernetes");
            InKubernetes = CheckInKubernetes();

            if (!InKubernetes) return;

            Logger.Log(KubernetesLog, "In Kubernetes!");
            jsonSettings = new JsonSerializerSettings {
                MissingMemberHandling = MissingMemberHandling.Ignore,
            };

            LoadKubeCA();
            Hostname = Dns.GetHostName();
            Namespace = MyNamespace();
            KubeToken = LoadKubeToken();
            var mTask = MySelf();
            mTask.Wait();
            Me = mTask.Result;
        }

        private static bool CheckInKubernetes() {
            try {
                return File.Exists($"{ServiceAccountPath}/token");
            } catch (Exception e) {
                Logger.Warn(KubernetesLog, $"Probably not in Kubernetes Mode: {e.Message}");
                return false;
            }
        }

        private static string MyNamespace() {
            return File.ReadAllText($"{ServiceAccountPath}/namespace");
        }

        private static string LoadKubeToken() {
            return File.ReadAllText($"{ServiceAccountPath}/token");
        }

        // Only for Kubernetes
        static bool KubernetesSslCheck(object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors policyErrors) {
            if (Configuration.IgnoreKubernetesCA) {
                return true;
            }
            Logger.Debug(KubernetesLog, "Validating Custom CA");
            foreach (var element in chain.ChainElements) {
                Logger.Debug(KubernetesLog, $"{element.Certificate.GetCertHashString()} == {KubeCAX509.GetCertHashString()}");
                if (element.Certificate.GetCertHashString().Equals(KubeCAX509.GetCertHashString())) {
                    return true;
                }
            }

            Logger.Debug(KubernetesLog, "Could not find any valid trusted certs");
            
            return false;
        }
        
        private static void LoadKubeCA() {
            KubeCAX509 = new X509Certificate($"{ServiceAccountPath}/ca.crt");
            Logger.Log(KubernetesLog, $"Loaded Kube CA: {KubeCAX509.GetCertHashString()}");
            ServicePointManager.ServerCertificateValidationCallback += KubernetesSslCheck;
        }

        private static async Task<Pod> MySelf() {
            var data = await Tools.Get($"{PodURL}/{Hostname}", KubeToken);
            return JsonConvert.DeserializeObject<Pod>(data, jsonSettings);
        }

        public static async Task<Service> Services() {
            var data = await Tools.Get(ServiceURL, KubeToken);
            return JsonConvert.DeserializeObject<Service>(data, jsonSettings);
        }

        public static async Task<Pods> Pods() {
            var data = await Tools.Get(PodURL, KubeToken);
            return JsonConvert.DeserializeObject<Pods>(data, jsonSettings);
        }
    }
}

File: kubernetes/kubernetes.go
package kubernetes

import (
	"encoding/json"
	"fmt"
	"github.com/mewkiz/pkg/osutil"
	"github.com/quan-to/remote-signer/models/KubernetesModels"
	"github.com/quan-to/slog"
	"io/ioutil"
	"os"
	"path"
)

const ServiceAccountPath = "/run/secrets/kubernetes.io/serviceaccount"

var inKubernetes = false
var currentNamespace = ""
var currentHostname = ""
var currentKubeToken = ""
var kubeLog = slog.Scope("Kubernetes")
var me *KubernetesModels.Pod

func serviceURL() string {
	return fmt.Sprintf("https://kubernetes.default.svc/api/v1/namespaces/%s/services", currentNamespace)
}

func podURL() string {
	return fmt.Sprintf("https://kubernetes.default.svc/api/v1/namespaces/%s/pods", currentNamespace)
}

func init() {
	setup()
}

func setup() {
	var err error
	kubeLog.Info("Checking if running in kubernetes")
	inKubernetes = checkInKubernetes()

	if !inKubernetes {
		kubeLog.Info("Not running in kubernetes...")
		return
	}

	kubeLog.Info("In Kubernetes!")
	currentHostname, err = os.Hostname()
	if err != nil {
		kubeLog.Error("Error getting hostname: %s", err)
	}

	currentNamespace = myNamespace()
	currentKubeToken = kubeToken()
	me = MySelf()
}

func checkInKubernetes() bool {
	return osutil.Exists(path.Join(ServiceAccountPath, "token"))
}

func myNamespace() string {
	data, err := ioutil.ReadFile(path.Join(ServiceAccountPath, "namespace"))
	if err != nil {
		kubeLog.Error("Error loading namespace: %s", err)
	}

	return string(data)
}
func kubeToken() string {
	data, err := ioutil.ReadFile(path.Join(ServiceAccountPath, "token"))
	if err != nil {
		kubeLog.Error("Error loading token: %s", err)
	}

	return string(data)
}

func Hostname() string {
	return currentHostname
}

func Namespace() string {
	return currentNamespace
}

func MySelf() *KubernetesModels.Pod {
	data, err := getWithToken(fmt.Sprintf("%s/%s", podURL(), currentHostname), currentKubeToken)

	if err != nil {
		kubeLog.Error("Error fetching Myself: %s", err)
		return nil
	}

	var pod KubernetesModels.Pod

	err = json.Unmarshal([]byte(data), &pod)
	if err != nil {
		kubeLog.Error("Error deserializing: %s", err)
	}

	return &pod
}

func Pods() []KubernetesModels.Pod {
	data, err := getWithToken(podURL(), currentKubeToken)
	if err != nil {
		kubeLog.Error("Error fetching Pods: %s", err)
		return nil
	}

	var podList KubernetesModels.PodList

	err = json.Unmarshal([]byte(data), &podList)
	if err != nil {
		kubeLog.Error("Error deserializing: %s", err)
	}

	return podList.Items
}

func Services() []KubernetesModels.Service {
	data, err := getWithToken(serviceURL(), currentKubeToken)
	if err != nil {
		kubeLog.Error("Error fetching Services: %s", err)
		return nil
	}

	var pods []KubernetesModels.Service

	err = json.Unmarshal([]byte(data), &pods)
	if err != nil {
		kubeLog.Error("Error deserializing: %s", err)
	}

	return pods
}

func Me() KubernetesModels.Pod {
	if me == nil {
		return KubernetesModels.Pod{}
	}

	return *me
}

func InKubernetes() bool {
	return inKubernetes
}
